freedomofpress/securedrop
# Validates all prod vars prior to installation

## Status

Ready for review.

(Confirmed with manual testing that the presence of the `openssh-client` directory is sufficient indication that SSH dotfiles persistence is enabled. This PR is no longer marked "Work in progress.")

## Description of Changes

Validates all site-specific production variables prior to running the prod playbook. Designed to catch common misconfiguration errors _before_ the playbook is run, to prevent late-stage failures that require excessive troubleshooting. Since we're switching to Ansible v2 (#1146), we can leverage the `msg` parameter on the `assert` task, which allows us to provide custom failure messages for variable validation, so Admins know where the problem lies and what to change.

The new validation logic requires Tails 3 (#1551) or greater as host OS, and additionally checks that persistence is enabled, and specifically that SSH dotfiles persistence is enabled. There are also functional checks that confirm the specified SMTP relay is accessible. It also pulls in @heartsucker's GPG validation logic presented in #1259, to confirm that the GPG public key files and full fingerprints match.

At some point in the future, we may wish to port the validation logic to pure Python, which would allow us to loop over input prompts when values don't pass validation, rather than collecting all variables and analyzing them as a set. 

Fixes #749. Fixes #1257. Obsoletes and therefore closes #1028. Includes and therefore closes #1259. 

## Testing

Requires testing on hardware, with an Admin Workstation on Tails 3 (#1551) configured with persistence and SecureDrop-specific config.

1. Boot into the Admin Workstation, with persistence enabled.
2. Run `./configure` in the ansible-base directory (see #1762 for more info) and configure working prod-compatible vars for GPG keys and OSSEC email authentication.
3. Confirm that the validation logic passes.

Then, to check failure mode:

1. Edit `group_vars/all/site-specific` with junk values, e.g. alter a single character in one of the GPG fingerprints.
2. Rerun `./configure`.
3. Confirm that the failure message is informative.

## Deployment

Mostly useful for first-time Admins during install time, but since the validation logic is run as part of the prod playbook, the changes should also catch future misconfiguration if required vars are not declared. Previously, such failures would only occur during the playbook run, which due to the Tor over ATHS strategy, could occur 10 minutes or more after executing the playbook command. 

## Checklist

Requiring manual testing on hardware to confirm functionality.
